### PR TITLE
[ruby2cpg] Fix ImplicitRequirePass.

### PR DESCRIPTION
**Pull request by ml86**
- Fix method lookup via AST edges. The approach of looking up module
  methods via is not possible at the execution time of the pass because
  AST is not yet linked.
  I replaced this buy a fullname based regex lookup.
  We should likely just change the execution time to after AST linking.

- The other changes are just for better readability and debugability.

@DavidBakerEffendi Please double check this PR thoroughly and ping me for discussion of running the ImplicitRequirePass later.

### CODE REVIEW

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/passes/ImplicitRequirePass.scala [hunk 1]
```diff
@@ -40,37 +40,48 @@ class ImplicitRequirePass(cpg: Cpg, programSummary: RubyProgramSummary) extends
...
+    val typeDecl = cpg.typeDecl.fullName(part.fullName + ".*").l
+    typeDecl.inheritsFromTypeFullName.foreach(identifiersToMatch.append)
+
+    val methods = findMethodsViaAstChildren(part).toList
+    val calls   = methods.ast.isCall.toList
+    val identifiers = calls.flatMap {
+      case x if x.name == Operators.alloc =>
+        x.argument.isIdentifier.name
+      case x =>
+        x.receiver.fieldAccess.fieldIdentifier.canonicalName
```
Comment (maltek) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/passes/ImplicitRequirePass.scala:65]: `.fieldAccess` likely doesn't do what we want here. It returns any field access recurively found below the `receiver`.

We either want to do `.isCall.nameExact(Operators.fieldAccess)`, or finally add a new helper that isn't just meant for the REPL.

**Review by maltek — CHANGES_REQUESTED** [review]

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/passes/ImplicitRequirePass.scala (hunk 1 above)
Comment (ml86) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/passes/ImplicitRequirePass.scala:65]: Wtf..., thanks for catching this. This kind of accessors should be completely deleted...

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/passes/ImplicitRequirePass.scala [hunk 2]
```diff
@@ -40,37 +40,48 @@ class ImplicitRequirePass(cpg: Cpg, programSummary: RubyProgramSummary) extends
...
-      case _           => Iterator.empty
-    }
+    // TODO For now we have to go via the full name regex because the AST is not yet linked
+    // at the execution time of this pass.
+    // Iterator(module) ++ module.astChildren.flatMap {
+    //  case x: TypeDecl => x.method.flatMap(findMethodsViaAstChildren)
+    //  case x: Method   => Iterator(x) ++ x.astChildren.collectAll[Method].flatMap(findMethodsViaAstChildren)
+    //  case _           => Iterator.empty
+    // }
+    cpg.method.fullName(module.fullName + ".*")
```
[the remark this answers is not in the document]
Comment (ml86) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/passes/ImplicitRequirePass.scala]: Done

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/passes/ImplicitRequirePass.scala (hunk 1 above)
Comment (ml86) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/passes/ImplicitRequirePass.scala]: Done

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/passes/ImplicitRequirePass.scala [hunk 3]
```diff
@@ -40,37 +40,48 @@ class ImplicitRequirePass(cpg: Cpg, programSummary: RubyProgramSummary) extends
...
-      }
-      .map {
-        case fi: FieldIdentifier => fi -> programSummary.matchingTypes(fi.canonicalName)
-        case i: Identifier       => i  -> programSummary.matchingTypes(i.name)
-      }
-      .distinct
-      .foreach { case (identifier, rubyTypes) =>
+    val identifiersToMatch = mutable.ArrayBuffer.empty[String]
+
+    val typeDecl = cpg.typeDecl.fullName(part.fullName + ".*").l
```
[the remark this answers is not in the document]
Comment (ml86) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/passes/ImplicitRequirePass.scala]: Done

**Review by maltek — APPROVED** [review]

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/passes/ImplicitRequirePass.scala (hunk 1 above)
Comment (maltek) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/passes/ImplicitRequirePass.scala]: I wouldn't mind removing them. I think there are some (potentially buggy) uses in the joern code though, that'd need to be checked and adapted.

**Review by DavidBakerEffendi — COMMENTED** [review]
Looks fine to me, hopefully this doesn't hurt performance too much and if it does, we can look at changing method linking.

**Pull request merged**